api/server/sdk/watcher.go

```diff
@@ -0,0 +1,242 @@
+package sdk
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/pborman/uuid"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	volumeEventType = "Volume"
+)
+
+type WatcherServer struct {
+	volumeServer *VolumeServer
+
+	watchConnections map[string][]*watchConnection
+	volumeChannel    chan *api.Volume
+	sync.RWMutex
+}
+
+// Watch streams a list of events back to clients. Events are resources that can typically fetch with openstorage API, such as Volumes,
+// Nodes, Disks, etc.
+//
+// Implementation note: the flow of data starts from porx ETCD changes, then the data will be send to openstorage via a golang channel.
+// Once the event object arrives at openstorage, it will be redistributed to a list of watch connections via another set of channels.
+func (w *WatcherServer) Watch(req *api.SdkWatchRequest, stream api.OpenStorageWatch_WatchServer) error {
+	if req.GetVolumeEvent() != nil {
+		return w.volumeWatch(req.GetVolumeEvent(), stream)
+	}
+	return nil
```

Comment:
Should we return an error here instead since we were unable to parse the request and/or the object is unsupported?